keycloak-plugin/sasl-plugin/src/main/java/io/enmasse/keycloak/spi/K8sServiceAccountCredentialProvider.java

```diff
@@ -0,0 +1,141 @@
+import org.keycloak.credential.CredentialProvider;
+import org.keycloak.models.KeycloakSession;
+import org.keycloak.models.RealmModel;
+import org.keycloak.models.UserCredentialModel;
+import org.keycloak.models.UserModel;
+import org.keycloak.models.cache.CachedUserModel;
+import org.keycloak.models.cache.OnUserCache;
+
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.util.Collections;
+import java.util.Objects;
+import java.util.Set;
+
+public class K8sServiceAccountCredentialProvider implements CredentialProvider, CredentialInputValidator, OnUserCache, CredentialInputUpdater {
+
+    private static final Logger LOG = Logger.getLogger(K8sServiceAccountCredentialProvider.class);
+
+    public static final String ENMASSE_SERVICE_ACCOUNT_TYPE = "enmasse-service-account";
+
+    K8sServiceAccountCredentialProvider(KeycloakSession session) {
+    }
+
+    @Override
+    public boolean supportsCredentialType(String s) {
+        return ENMASSE_SERVICE_ACCOUNT_TYPE.equals(s);
+    }
+
+    @Override
+    public boolean isConfiguredFor(RealmModel realmModel, UserModel userModel, String s) {
+        return "serviceaccount".equals(userModel.getFirstAttribute("authenticationType")) ;
+    }
+
+    @Override
+    public boolean isValid(RealmModel realmModel, UserModel userModel, CredentialInput credentialInput) {
+        JsonObject body = new JsonObject();
+
+        body.put("kind", "TokenReview");
+        body.put("apiVersion", "authentication.k8s.io/v1beta1");
+
+        JsonObject spec = new JsonObject();
+        spec.put("token", ((UserCredentialModel)credentialInput).getValue());
+        body.put("spec", spec);
+
+        JsonObject responseBody= doRawHttpRequest("/apis/authentication.k8s.io/v1beta1/tokenreviews", "POST", body, false);
+
+        JsonObject status = responseBody.getJsonObject("status");
+
+        boolean authenticated = false;
+        String userName = null;
+        if (status != null) {
+            Boolean auth = status.getBoolean("authenticated");
+            authenticated = auth == null ? false : auth;
+            JsonObject user = status.getJsonObject("user");
+            if (user != null) {
+                userName = user.getString("username");
+            }
+        }
+        if(!authenticated) {
+            LOG.debug("User: " + userModel.getUsername() + " not authenticated for realm " + realmModel.getName());
+        } else if(!Objects.equals(userModel.getUsername(), userName)) {
+            LOG.debug("Attempt to log in for user " + userModel.getUsername() + " in realm " + realmModel.getName() + " with token for " + userName);
+        }
+
+        return authenticated && Objects.equals(userModel.getUsername(), userName);
+    }
+
+    private JsonObject doRawHttpRequest(String path, String method, JsonObject body, boolean errorOk) {
+        NamespacedOpenShiftClient client = new DefaultOpenShiftClient();
+        OkHttpClient httpClient = client.adapt(OkHttpClient.class);
```

Comment:
(If we want to use the kubernetes-client etc.) I think you need to close both of these to avoid a resource leak

Reply:
I literally just copied this code from api-server :-)

Reply:
The new DefaultOpenshiftClient() is the one leaking. In the api-server it is created at the start of the server process.

Reply:
I've moved the creation of the DefaultOpenShiftClient to the constructor and closed it in the close method of the provider.